api/src/services/payload.ts

```diff
@@ -190,16 +190,12 @@ export class PayloadService {
 			});
 		}
 
-		await Promise.all(
-			processedPayload.map(async (record: any) => {
-				await Promise.all(
-					specialFields.map(async ([name, field]) => {
-						const newValue = await this.processField(field, record, action, this.accountability);
-						if (newValue !== undefined) record[name] = newValue;
-					}),
-				);
-			}),
-		);
+		for (const record of processedPayload) {
+			for (const [name, field] of specialFields) {
+				const newValue = await this.processField(field, record, action, this.accountability);
+				if (newValue !== undefined) record[name] = newValue;
+			}
+		}
```

Comment:
This now runs sequential rather than parallel. Did that help the performance increase? I take it that at the tens of thousands of items range the number of async handles becomes a new bottleneck?

Reply:
Actually, due to the mono-thread aspect of JS, the Node's event loop can't do real parallel execution here (in contrast with calling multiple endpoints from a web browser for example).
As a result it's just flooding the event loop Promises callstack producing a slowdown instead of a faster run.

The Primagen has done a very clear video about it: https://youtu.be/dWxcw2R0gZs?t=480

Reply:
Ah yes processField isn't touching the DB of course 🙂